homeassistant/components/entur_public_transport/sensor.py

```diff
@@ -170,33 +195,51 @@ def icon(self) -> str:
         """Icon to use in the frontend."""
         return self._icon
 
-    def update(self) -> None:
+    async def async_update(self) -> None:
         """Get the latest data and update the states."""
-        from enturclient.consts import (
-            ATTR, ATTR_EXPECTED_AT, ATTR_NEXT_UP_AT, CONF_LOCATION,
-            CONF_LATITUDE as LAT, CONF_LONGITUDE as LONG, CONF_TRANSPORT_MODE)
-
-        self.api.update()
+        await self.api.async_update()
 
         data = self.api.get_stop_info(self._stop)
-        if data is not None and ATTR in data:
-            attrs = data[ATTR]
-            self._attributes.update(attrs)
-
-            if ATTR_NEXT_UP_AT in attrs:
-                self._attributes[ATTR_NEXT_UP_IN] = \
-                    due_in_minutes(attrs[ATTR_NEXT_UP_AT])
-
-            if CONF_LOCATION in data and self._show_on_map:
-                self._attributes[CONF_LATITUDE] = data[CONF_LOCATION][LAT]
-                self._attributes[CONF_LONGITUDE] = data[CONF_LOCATION][LONG]
-
-            if ATTR_EXPECTED_AT in attrs:
-                self._state = due_in_minutes(attrs[ATTR_EXPECTED_AT])
-            else:
-                self._state = None
-
-            self._icon = ICONS.get(
-                data[CONF_TRANSPORT_MODE], ICONS[DEFAULT_ICON_KEY])
-        else:
+
+        if data is None:
+            self._state = None
+            return
+
+        if self._show_on_map and data.latitude and data.longitude:
+            self._attributes[CONF_LATITUDE] = data.latitude
+            self._attributes[CONF_LONGITUDE] = data.longitude
+
+        calls = data.estimated_calls
+        if not calls:
             self._state = None
+            return
+
+        self._state = due_in_minutes(calls[0].expected_departure_time)
+        self._icon = ICONS.get(
+            calls[0].transport_mode, ICONS[DEFAULT_ICON_KEY])
+        self._attributes[ATTR_STOP_ID] = data.place_id
+
+        self._attributes[ATTR_ROUTE] = calls[0].front_display
+        self._attributes[ATTR_ROUTE_ID] = calls[0].line_id
+        self._attributes[ATTR_EXPECTED_AT] = calls[0]\
+            .expected_departure_time.strftime("%H:%M")
+        self._attributes[ATTR_REALTIME] = calls[0].is_realtime
+        self._attributes[ATTR_DELAY] = calls[0].delay_in_min
+
+        if len(calls) > 1:
```

Comment:
Maybe this is cleaner:

 if len(calls) < 2:
   return

Reply:
Yeah, agree. That would reduce a lot of the nesting. Have updated it, and look a bit more smooth now.